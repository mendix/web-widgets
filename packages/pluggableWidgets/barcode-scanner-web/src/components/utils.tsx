import {
    BinaryBitmap,
    BrowserMultiFormatReader,
    HTMLCanvasElementLuminanceSource,
    HybridBinarizer,
    BarcodeFormat,
    DecodeHintType,
    Result
} from "@zxing/library";
import { BarcodeFormatsType } from "typings/BarcodeScannerProps";

export const returnVideoWidthHeight = (
    curVideoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): { width: number; height: number } => {
    const aspectRatioClient = curVideoRef.clientWidth / curVideoRef.clientHeight;
    const aspectRatioVideo = curVideoRef.videoWidth / curVideoRef.videoHeight;

    let videoCropWidth = (canvasMiddle.clientWidth / curVideoRef.clientWidth) * curVideoRef.videoWidth;
    let videoCropHeight = (canvasMiddle.clientHeight / curVideoRef.clientHeight) * curVideoRef.videoHeight;

    if (aspectRatioVideo < aspectRatioClient) {
        videoCropHeight = videoCropWidth;
    } else {
        videoCropWidth = videoCropHeight;
    }
    return { width: videoCropWidth, height: videoCropHeight };
};

export const drawCropOnCanvas = (
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): HTMLCanvasElement => {
    const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
    const videoWidthHeight = returnVideoWidthHeight(videoRef, canvasMiddle);
    if (
        Math.abs(captureCanvas.width - videoWidthHeight.width) > 1 ||
        Math.abs(captureCanvas.height - videoWidthHeight.height) > 1
    ) {
        captureCanvas.width = videoWidthHeight.width;
        captureCanvas.height = videoWidthHeight.height;
    }
    if (canvasContext !== null && videoRef !== null) {
        canvasContext.drawImage(
            videoRef,
            (videoRef.videoWidth - videoWidthHeight.width) / 2,
            (videoRef.videoHeight - videoWidthHeight.height) / 2,
            videoWidthHeight.width,
            videoWidthHeight.height,
            0,
            0,
            videoWidthHeight.width,
            videoWidthHeight.height
        );
    }
    return captureCanvas;
};

export function decodeBarcodeDetector(
    resolve: (value: Result | string) => void,
    reject: (reason?: Error) => void,
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement,
    barcodeDetector?: any
): void {
    const croppedOnCanvas = drawCropOnCanvas(captureCanvas, videoRef, canvasMiddle);
    barcodeDetector
        .detect(croppedOnCanvas)
        .then((barcodes: any) => {
            if (barcodes[0]?.rawValue) {
                resolve(barcodes[0]?.rawValue);
            } else {
                setTimeout(
                    () =>
                        decodeBarcodeDetector(resolve, reject, captureCanvas, videoRef, canvasMiddle, barcodeDetector),
                    50
                );
            }
        })
        .catch((err: any) => {
            reject(err);
        });
}

export const decodeCanvas = (reader: BrowserMultiFormatReader, captureCanvas: HTMLCanvasElement): Result => {
    const luminanceSource = new HTMLCanvasElementLuminanceSource(captureCanvas);

    const binaryBitmap = new BinaryBitmap(new HybridBinarizer(luminanceSource));
    const result = reader.decodeBitmap(binaryBitmap);
    return result;
};

export const createHints = (
    useAllFormats: boolean,
    barcodeFormats?: BarcodeFormatsType[]
): Map<DecodeHintType, any> => {
    const hints = new Map();
    let formats: BarcodeFormat[] = [];
    if (useAllFormats) {
        formats = [
            BarcodeFormat.UPC_A,
            BarcodeFormat.UPC_E,
            BarcodeFormat.EAN_8,
            BarcodeFormat.EAN_13,
            BarcodeFormat.CODE_39,
            BarcodeFormat.CODE_128,
            BarcodeFormat.ITF,
            BarcodeFormat.RSS_14,
            BarcodeFormat.QR_CODE,
            BarcodeFormat.DATA_MATRIX,
            BarcodeFormat.AZTEC,
            BarcodeFormat.PDF_417
        ];
    } else {
        if (barcodeFormats) {
            formats = barcodeFormats.map(val => BarcodeFormat[val.barcodeFormat]);
        }
    }
    hints.set(DecodeHintType.POSSIBLE_FORMATS, formats);
    hints.set(DecodeHintType.ENABLE_CODE_39_EXTENDED_MODE, true);
    return hints;
};
