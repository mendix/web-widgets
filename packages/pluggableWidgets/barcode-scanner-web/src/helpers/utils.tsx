import {
    BarcodeFormat,
    BinaryBitmap,
    BrowserMultiFormatReader,
    DecodeHintType,
    HTMLCanvasElementLuminanceSource,
    HybridBinarizer,
    Result
} from "@zxing/library";
import { RefObject } from "react";
import { BarcodeFormatsType } from "typings/BarcodeScannerProps";

export const mediaStreamConstraints: MediaStreamConstraints = {
    audio: false,
    video: {
        facingMode: "environment",
        width: { min: 1280, ideal: 4096, max: 4096 },
        height: { min: 720, ideal: 2160, max: 2160 }
    }
};

export type ReaderProps = {
    onSuccess?: (data: string) => void;
    onError?: (e: Error) => void;
    useCrop: boolean;
    barcodeFormats?: BarcodeFormatsType[];
    useAllFormats: boolean;
    canvasMiddleRef: RefObject<HTMLDivElement>;
    detectionLogic?: "zxing" | "native";
};

export type UseReaderHook = (args: ReaderProps) => { ref: RefObject<HTMLVideoElement>; useBrowserAPI: boolean };

export const returnVideoWidthHeight = (
    curVideoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): { width: number; height: number } => {
    const aspectRatioClient = curVideoRef.clientWidth / curVideoRef.clientHeight;
    const aspectRatioVideo = curVideoRef.videoWidth / curVideoRef.videoHeight;

    let videoCropWidth = (canvasMiddle.clientWidth / curVideoRef.clientWidth) * curVideoRef.videoWidth;
    let videoCropHeight = (canvasMiddle.clientHeight / curVideoRef.clientHeight) * curVideoRef.videoHeight;

    if (aspectRatioVideo < aspectRatioClient) {
        videoCropHeight = videoCropWidth;
    } else {
        videoCropWidth = videoCropHeight;
    }
    return { width: videoCropWidth, height: videoCropHeight };
};

export const drawCropOnCanvas = (
    captureCanvas: HTMLCanvasElement,
    videoRef: HTMLVideoElement,
    canvasMiddle: HTMLDivElement
): HTMLCanvasElement => {
    const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
    const videoWidthHeight = returnVideoWidthHeight(videoRef, canvasMiddle);
    if (
        Math.abs(captureCanvas.width - videoWidthHeight.width) > 1 ||
        Math.abs(captureCanvas.height - videoWidthHeight.height) > 1
    ) {
        captureCanvas.width = videoWidthHeight.width;
        captureCanvas.height = videoWidthHeight.height;
    }
    if (canvasContext !== null && videoRef !== null) {
        canvasContext.drawImage(
            videoRef,
            (videoRef.videoWidth - videoWidthHeight.width) / 2,
            (videoRef.videoHeight - videoWidthHeight.height) / 2,
            videoWidthHeight.width,
            videoWidthHeight.height,
            0,
            0,
            videoWidthHeight.width,
            videoWidthHeight.height
        );
    }
    return captureCanvas;
};

export const decodeCanvas = (reader: BrowserMultiFormatReader, captureCanvas: HTMLCanvasElement): Result => {
    const luminanceSource = new HTMLCanvasElementLuminanceSource(captureCanvas);

    const binaryBitmap = new BinaryBitmap(new HybridBinarizer(luminanceSource));
    const result = reader.decodeBitmap(binaryBitmap);
    return result;
};

export const createHints = (
    useAllFormats: boolean,
    barcodeFormats?: BarcodeFormatsType[]
): Map<DecodeHintType, any> => {
    const hints = new Map();
    let formats: BarcodeFormat[] = [];
    if (useAllFormats) {
        formats = [
            BarcodeFormat.UPC_A,
            BarcodeFormat.UPC_E,
            BarcodeFormat.EAN_8,
            BarcodeFormat.EAN_13,
            BarcodeFormat.CODE_39,
            BarcodeFormat.CODE_128,
            BarcodeFormat.ITF,
            BarcodeFormat.RSS_14,
            BarcodeFormat.QR_CODE,
            BarcodeFormat.DATA_MATRIX,
            BarcodeFormat.AZTEC,
            BarcodeFormat.PDF_417
        ];
    } else {
        if (barcodeFormats) {
            formats = barcodeFormats.map(val => BarcodeFormat[val.barcodeFormat]);
        }
    }
    hints.set(DecodeHintType.POSSIBLE_FORMATS, formats);
    hints.set(DecodeHintType.ENABLE_CODE_39_EXTENDED_MODE, true);
    return hints;
};
