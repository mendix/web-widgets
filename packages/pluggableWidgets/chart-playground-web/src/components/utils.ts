function isNonEmptyString(value: undefined | null | string): value is string {
    return value !== null && value !== undefined && value !== "";
}

export const emptyObjectString = "{}";

export function ifNonEmptyStringElseEmptyObjectString(value: undefined | null | string): string {
    return isNonEmptyString(value) ? value : emptyObjectString;
}
