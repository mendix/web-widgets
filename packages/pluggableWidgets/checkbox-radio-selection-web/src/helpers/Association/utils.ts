import {
    ActionValue,
    DynamicValue,
    ListAttributeValue,
    ListExpressionValue,
    ListValue,
    ListWidgetValue,
    ReferenceSetValue,
    ReferenceValue,
    ValueStatus
} from "mendix";
import {
    CheckboxRadioSelectionContainerProps,
    OptionsSourceCustomContentTypeEnum
} from "../../../typings/CheckboxRadioSelectionProps";

export function extractAssociationProps(
    props: CheckboxRadioSelectionContainerProps
): [
    ReferenceValue | ReferenceSetValue,
    ListValue,
    ListAttributeValue<string> | ListExpressionValue<string>,
    DynamicValue<string>,
    boolean,
    ActionValue | undefined,
    ListWidgetValue | undefined,
    OptionsSourceCustomContentTypeEnum
] {
    const attr = props.attributeAssociation;
    const ds = props.optionsSourceAssociationDataSource!;

    // Determine caption provider based on caption type
    const captionProvider =
        props.optionsSourceAssociationCaptionType === "attribute"
            ? props.optionsSourceAssociationCaptionAttribute!
            : props.optionsSourceAssociationCaptionExpression!;

    // For simplicity, we'll create a basic empty option
    const emptyOption: DynamicValue<string> = {
        status: ValueStatus.Available,
        value: ""
    };

    // Checkbox Radio Selection controls don't need clearable like combobox does
    const clearable = false;

    const onChangeEvent = props.onChangeEvent;
    const customContent = props.optionsSourceAssociationCustomContent;
    const customContentType = props.optionsSourceCustomContentType;

    return [attr, ds, captionProvider, emptyOption, clearable, onChangeEvent, customContent, customContentType];
}
