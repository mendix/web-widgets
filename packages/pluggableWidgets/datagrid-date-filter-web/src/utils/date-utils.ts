/**
 * Change the time of a date and return an UTC date
 * @param date
 * @param hours
 * @param minutes
 * @param seconds
 */
export function changeTimeToMidnight(date: Date): Date {
    const newDate = new Date(date.getTime());
    newDate.setHours(0);
    newDate.setMinutes(0);
    newDate.setSeconds(0);
    newDate.setMilliseconds(0);
    return newDate;
}

/**
 * This function takes format string and replace
 * single "d" and "M" with their "double" equivalent
 * Main purpose is to "fix" value formats accepted
 * by DateFilter input. Idea is that if format string is
 * uses short format (e.g d-M-yyyy) user should still
 * be able to type value with leading zeros "09-02-2002".
 *
 * Example:
 * "d/M/yyyy" -> "dd/MM/yyyy"
 *
 * @param formatString
 */
export function doubleMonthOrDayWhenSingle(formatString: string): string {
    return formatString.replaceAll(/d+|M+/g, m => (m.length > 1 ? m : m + m));
}

/**
 * Given a date formatter string replaces all uppercase 'E' to the lowercase 'e'
 * this function returns the same string with the above characters replaced.
 * Example: "YYww.E" returns "YYww.e"
 * @function dayOfWeekWhenUpperCase
 * @param {string} formatString The string used to find cases of uppercase E.
 * @return {string} The same string but with 'E' in lowercase ('e').
 *
 */
export function dayOfWeekWhenUpperCase(formatString: string): string {
    return formatString.replaceAll(/E/g, m => m.toLowerCase());
}
