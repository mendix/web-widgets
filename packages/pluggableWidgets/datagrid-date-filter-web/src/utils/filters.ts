import { InitialFilterValue } from "@mendix/widget-plugin-filtering";
import { DefaultFilterEnum } from "../../typings/DatagridDateFilterProps";

export type DefaultFilterValue = {
    type: DefaultFilterEnum;
    value?: Date;
    startDate?: Date;
    endDate?: Date;
};

export function translateFilters(filters?: InitialFilterValue[]): DefaultFilterValue | undefined {
    if (filters && filters.length > 0) {
        if (filters.length === 1) {
            const [filter] = filters;
            let type: DefaultFilterEnum = "greater";
            switch (filter.type) {
                case ">":
                    type = "greater";
                    break;
                case ">=":
                    type = "greaterEqual";
                    break;
                case "<":
                    type = "smaller";
                    break;
                case "<=":
                    type = "smallerEqual";
                    break;
            }

            return {
                type,
                value: ensureDate(filter.value)
            };
        } else {
            const [filterStart, filterEnd] = filters;
            let type: DefaultFilterEnum = "equal";
            if (filterStart.type === ">=" && filterEnd.type === "<") {
                type = "equal";
            } else if (filterStart.type === "<" && filterEnd.type === ">=") {
                type = "notEqual";
            } else if (filterStart.type === ">=" && filterEnd.type === "<=") {
                return {
                    type: "between",
                    startDate: ensureDate(filterStart.value),
                    endDate: ensureDate(filterEnd.value)
                };
            }
            return {
                type,
                value: ensureDate(filterStart.value)
            };
        }
    }
    return undefined;
}

function ensureDate(value: unknown): Date {
    if (!(value instanceof Date)) {
        throw Error("Invalid init filter value");
    }
    return value;
}
