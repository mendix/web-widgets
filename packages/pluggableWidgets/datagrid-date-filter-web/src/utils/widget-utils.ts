import { DatagridDateFilterContainerProps } from "../../typings/DatagridDateFilterProps";

export function isLoadingDefaultValues(props: DatagridDateFilterContainerProps): boolean {
    const statusList = [props.defaultValue?.status, props.defaultStartDate?.status, props.defaultEndDate?.status];
    return statusList.some(status => status === "loading");
}
