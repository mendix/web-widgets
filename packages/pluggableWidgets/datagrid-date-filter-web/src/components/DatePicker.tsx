import {
    createElement,
    Dispatch,
    forwardRef,
    Fragment,
    MutableRefObject,
    ReactElement,
    SetStateAction,
    useCallback,
    useMemo,
    useRef,
    useState
} from "react";
import DatePickerComponent from "react-datepicker";
import classNames from "classnames";
import { isDate, isValid } from "date-fns";
import replaceAllInserter from "string.prototype.replaceall";
import { doubleMonthOrDayWhenSingle } from "../utils/utils";
import CalendarIcon from "./CalendarIcon";

export type RangeDateValue = [Date | undefined, Date | undefined];

interface DatePickerProps {
    adjustable: boolean;
    dateFormat?: string;
    disabledInput?: boolean;
    enableRange?: boolean;
    locale?: string;
    id?: string;
    placeholder?: string;
    setRangeValues?: Dispatch<SetStateAction<RangeDateValue>>;
    rangeValues?: RangeDateValue;
    screenReaderCalendarCaption?: string;
    screenReaderInputCaption?: string;
    setValue: Dispatch<SetStateAction<Date | undefined>>;
    value?: Date;
    calendarStartDay?: number;
}

replaceAllInserter.shim();

export const DatePicker = forwardRef(
    (props: DatePickerProps, ref: MutableRefObject<DatePickerComponent> | null): ReactElement => {
        const [open, setOpen] = useState(false);
        const buttonRef = useRef<HTMLButtonElement>(null);
        const dateFilterContainerRef = useRef<HTMLDivElement>(null);
        const id = useMemo(() => `datepicker_` + Math.random(), []);

        const buttonClick = useCallback(() => {
            setOpen(open => !open);
            setTimeout(() => {
                (dateFilterContainerRef.current?.querySelector(".react-datepicker > button") as HTMLElement)?.focus();
            }, 10);
        }, []);

        let dateFormats;
        if (props.dateFormat) {
            // Replace with full patterns d -> dd, M -> MM
            const fixedFormatString = doubleMonthOrDayWhenSingle(props.dateFormat);
            dateFormats =
                props.dateFormat !== fixedFormatString ? [props.dateFormat, fixedFormatString] : [props.dateFormat];
        }

        return (
            <Fragment>
                <div ref={dateFilterContainerRef} id={id} className="date-filter-container" />
                <span className="sr-only" id={`${props.id}-label`}>
                    {props.screenReaderInputCaption}
                </span>
                <DatePickerComponent
                    allowSameDay={false}
                    ariaLabelledBy={`${props.id}-label`}
                    autoFocus={false}
                    calendarStartDay={props.calendarStartDay}
                    className={classNames("form-control", { "filter-input": props.adjustable })}
                    dateFormat={dateFormats}
                    disabled={props.disabledInput ?? false}
                    disabledKeyboardNavigation={false}
                    dropdownMode="select"
                    enableTabLoop
                    endDate={props.enableRange ? props.rangeValues?.[1] : undefined}
                    isClearable={props.enableRange}
                    locale={props.locale}
                    onChange={date => {
                        if (Array.isArray(date)) {
                            const [startDate, endDate] = date;
                            props.setRangeValues?.([startDate ?? undefined, endDate ?? undefined]);
                            props.setValue(undefined);
                        } else {
                            if (isDate(date) && isValid(date)) {
                                props.setValue(date as Date);
                            } else {
                                props.setValue(undefined);
                            }
                            props.setRangeValues?.([undefined, undefined]);
                        }
                    }}
                    onClickOutside={event => {
                        if (!buttonRef.current || buttonRef.current.contains(event.target as Node)) {
                            return;
                        }
                        setOpen(false);
                    }}
                    open={open}
                    onInputClick={() => setOpen(true)}
                    placeholderText={props.placeholder}
                    popperPlacement="bottom-end"
                    popperModifiers={[
                        {
                            name: "offset",
                            options: {
                                offset: [0, 0]
                            }
                        }
                    ]}
                    portalId={id}
                    preventOpenOnFocus
                    readOnly={props.enableRange}
                    ref={ref}
                    startDate={props.enableRange ? props.rangeValues?.[0] : undefined}
                    selected={props.enableRange ? undefined : props.value}
                    selectsRange={props.enableRange}
                    shouldCloseOnSelect={false}
                    showMonthDropdown
                    showPopperArrow={false}
                    showYearDropdown
                    strictParsing
                    useWeekdaysShort={false}
                />
                <button
                    aria-controls={id}
                    aria-expanded={open}
                    aria-haspopup
                    aria-label={props.screenReaderCalendarCaption ?? "Show calendar"}
                    ref={buttonRef}
                    className="btn btn-default btn-calendar"
                    onClick={buttonClick}
                    onKeyDown={e => {
                        if (e.key === "Enter" || e.key === " ") {
                            e.preventDefault();
                            e.stopPropagation();
                            buttonClick();
                        }
                    }}
                >
                    <CalendarIcon />
                </button>
            </Fragment>
        );
    }
);
