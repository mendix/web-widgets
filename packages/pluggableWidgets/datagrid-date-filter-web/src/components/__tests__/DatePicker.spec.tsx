import { createElement } from "react";
import { render as renderEnzyme } from "enzyme";
import { fireEvent, render } from "@testing-library/react";
import { DatePicker } from "../DatePicker";
import ReactDOM from "react-dom";
import { doubleMonthOrDayWhenSingle } from "../../utils/date-utils";

describe("Date picker component", () => {
    beforeAll(() => {
        jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);

        jest.spyOn(ReactDOM, "createPortal").mockImplementation((element, _node, _key) => {
            return element as ReturnType<typeof ReactDOM.createPortal>;
        });
    });

    it("renders correctly", () => {
        const component = renderEnzyme(
            <DatePicker adjustable setValue={jest.fn()} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly when is not adjustable", () => {
        const component = renderEnzyme(
            <DatePicker adjustable={false} setValue={jest.fn()} dateFormat="dd/MM/yyyy" locale="nl-NL" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with different locale and date format", () => {
        const component = renderEnzyme(
            <DatePicker adjustable={false} setValue={jest.fn()} dateFormat="yyyy-MM-dd" locale="pt-BR" />
        );

        expect(component).toMatchSnapshot();
    });

    it("renders correctly with a11y properties", () => {
        const component = renderEnzyme(
            <DatePicker
                adjustable
                setValue={jest.fn()}
                dateFormat="yyyy-MM-dd"
                locale="pt-BR"
                screenReaderInputCaption="my input"
                screenReaderCalendarCaption="my calendar"
            />
        );

        expect(component).toMatchSnapshot();
    });

    it("calls for setValue when value changes", async () => {
        const setValue = jest.fn();
        const component = render(
            <DatePicker
                adjustable
                setValue={setValue}
                dateFormat="dd/MM/yyyy"
                locale="nl-NL"
                placeholder="Placeholder"
            />
        );

        fireEvent.change(component.getByPlaceholderText("Placeholder"), { target: { value: "01/12/2020" } });

        expect(setValue).toBeCalledTimes(1);
    });

    it("calls for setRangeValues when value changes", async () => {
        const setRangeValues = jest.fn();
        const component = render(
            <DatePicker
                adjustable
                setValue={jest.fn()}
                setRangeValues={setRangeValues}
                dateFormat="dd/MM/yyyy"
                enableRange
                locale="nl-NL"
                placeholder="Placeholder"
            />
        );

        fireEvent.change(component.getByPlaceholderText("Placeholder"), {
            // Trick to trigger events (Should be dd/MM/yyyy - dd/MM/yyyy) but the library does not validate values coming from the input. That's why it is readonly when running on the browser.
            target: { value: "01/12/2020" }
        });
        expect(setRangeValues).toBeCalledTimes(1);
    });

    test.each([
        ["d/m/yyyy", "dd/m/yyyy"],
        ["dd/m/yyyy", "dd/m/yyyy"],
        ["d/MM/yyyy", "dd/MM/yyyy"],
        ["d-m-yyyy", "dd-m-yyyy"],
        ["d-mm-yyyy", "dd-mm-yyyy"],
        ["d-M-yyyy", "dd-MM-yyyy"],
        ["d-MM-yyyy", "dd-MM-yyyy"],
        ["dd-m-yyyy", "dd-m-yyyy"],
        ["dd-mm-yyyy", "dd-mm-yyyy"],
        ["dd-M-yyyy", "dd-MM-yyyy"],
        ["dd-MM-yyyy", "dd-MM-yyyy"],
        ["dd/MM/yyyy", "dd/MM/yyyy"],
        ["d/MM/yyyy", "dd/MM/yyyy"],
        ["d/M/yyyy", "dd/MM/yyyy"],
        ["dd/M/yyyy", "dd/MM/yyyy"],
        ["dd/m/yyyy", "dd/m/yyyy"],
        ["dd/mm/yyyy", "dd/mm/yyyy"],
        ["yyyy-MM-dd", "yyyy-MM-dd"],
        ["yyyy-MM-d", "yyyy-MM-dd"],
        ["yyyy-m-d", "yyyy-m-dd"],
        ["yyyy-mm-dd", "yyyy-mm-dd"],
        ["yyyy-mm-d", "yyyy-mm-dd"],
        ["yyyy.MM.dd", "yyyy.MM.dd"],
        ["yyyy.M.d", "yyyy.MM.dd"],
        ["yyyy.MM.d", "yyyy.MM.dd"],
        ["yyyy.m.dd", "yyyy.m.dd"],
        ["yyyy.M.dd", "yyyy.MM.dd"],
        ["yyyy.mm.dd", "yyyy.mm.dd"],
        ["yyyy.mm.d", "yyyy.mm.dd"],
        ["yyyyMMdd", "yyyyMMdd"],
        ["yyyyMdd", "yyyyMMdd"],
        ["dMMyyyy", "ddMMyyyy"],
        ["ddMMyyyy", "ddMMyyyy"],
        ["ddMyyyy", "ddMMyyyy"]
    ])("expect for input %p output to be %p", (input, expected) => {
        expect(doubleMonthOrDayWhenSingle(input)).toStrictEqual(expected);
    });
});
