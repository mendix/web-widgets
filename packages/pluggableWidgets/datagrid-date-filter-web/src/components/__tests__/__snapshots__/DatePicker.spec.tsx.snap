// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date picker component renders correctly 1`] = `
[
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />,
  <span
    class="sr-only"
    id="undefined-label"
  />,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control filter-input"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Show calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>,
]
`;

exports[`Date picker component renders correctly when is not adjustable 1`] = `
[
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />,
  <span
    class="sr-only"
    id="undefined-label"
  />,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Show calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>,
]
`;

exports[`Date picker component renders correctly with a11y properties 1`] = `
[
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />,
  <span
    class="sr-only"
    id="undefined-label"
  >
    my input
  </span>,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control filter-input"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="my calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>,
]
`;

exports[`Date picker component renders correctly with different locale and date format 1`] = `
[
  <div
    class="date-filter-container"
    id="datepicker_0.123456789"
  />,
  <span
    class="sr-only"
    id="undefined-label"
  />,
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container "
    >
      <span
        aria-live="polite"
        class="react-datepicker__aria-live"
        role="alert"
      />
      <input
        aria-labelledby="undefined-label"
        class="form-control"
        type="text"
        value=""
      />
    </div>
  </div>,
  <button
    aria-controls="datepicker_0.123456789"
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Show calendar"
    class="btn btn-default btn-calendar"
  >
    <svg
      class="button-icon"
      viewBox="0 0 32 32"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M27,5h-3V3h-4v2H12V3h-4v2h-3c-.55,0-1,.45-1,1V28c0,.55,.45,1,1,1H27c.55,0,1-.45,1-1V6c0-.55-.45-1-1-1ZM12,23h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,5h-4v-2h4v2Zm0-5h-4v-2h4v2Zm6,0h-4v-2h4v2Zm0-5H8v-2H24v2Z"
        fill="currentColor"
      />
    </svg>
  </button>,
]
`;
